### PR TITLE
Golang Support

### PR DESCRIPTION
**Pull request by threeguts**
## Summary

Added support for Golang

## Validation

- [X] I ran the relevant tests locally.
- [X] I included the executed commands and outcomes in the PR description.

## TDD RED Evidence Gate (when applicable)

Reference policy: `docs/tdd-red-evidence.md`

- [ ] This PR does **not** require RED evidence (explain why), or
- [ ] RED evidence exists on all required Jira test subtasks before implementation completion.
- [ ] RED evidence comments include command, failing result, and meaningful failure message.
- [ ] No tests were bypassed with `#[ignore]` or equivalent shortcuts.

## Jira

- [X] Linked Jira issue(s) are updated with implementation notes and test results.
- [X] Final status transitions match the task definition of done.

### CODE REVIEW

File: bitloops/src/adapters/languages/go/canonical.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+use crate::host::devql::CanonicalKindProjection;
+use crate::host::language_adapter::{CanonicalMapping, MappingCondition};
+
+pub(crate) static GO_CANONICAL_MAPPINGS: &[CanonicalMapping] = &[
+    CanonicalMapping {
+        language_kind: "function_declaration",
```
Comment (giorgosnty) [inline bitloops/src/adapters/languages/go/canonical.rs]: Not your mistake, I saw that in the previous implementations lamguage kind is magic strings, could you fix it at least in your PR and then improve the rest languages to use an enum?

File: bitloops/src/adapters/languages/go/canonical.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        language_kind: "type_spec",
+        projection: CanonicalKindProjection::Type,
+        condition: MappingCondition::Always,
+    },
+    CanonicalMapping {
+        language_kind: "type_alias",
+        projection: CanonicalKindProjection::Type,
+        condition: MappingCondition::Always,
+    },
+    CanonicalMapping {
```
Comment (giorgosnty) [inline bitloops/src/adapters/languages/go/canonical.rs:25]: Isn't struct type just type_spec? Is it different from tree sitter's perspective?

File: bitloops/src/adapters/languages/go/edges.rs [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+    for child in node.named_children(&mut cursor) {
+        if child.kind() == "type_elem"
+            && let Some(type_node) = child.named_child(0)
+        {
+            push_go_embedding_edge(owner.symbol_fqn.clone(), type_node, traversal);
+        }
+    }
+}
+
+fn push_go_embedding_edge(
```
Comment (giorgosnty) [inline bitloops/src/adapters/languages/go/edges.rs:393]: what do we mean embedding in this context?

File: bitloops/src/adapters/languages/go/edges.rs [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+        return format!("package::{package_name}::member::{field_name}");
+    }
+    if let Some(operand_name) = operand_name {
+        return format!("{path}::{operand_name}::{field_name}");
+    }
+    format!("{path}::member::{field_name}")
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (giorgosnty) [inline bitloops/src/adapters/languages/go/edges.rs]: Let's do some more exact testing e.g we have 4 edges 
1 -> imports
2 -> calls etc 

Also are there any edge cases where we have duplicate edges where we shouldn't? e.g. calls and references, do we need both?

File: bitloops/src/adapters/languages/go/canonical.rs (hunk 2 above)
Comment (threeguts) [reply bitloops/src/adapters/languages/go/canonical.rs:25]: It's different, type_spec is the named declaration and the struct_type is the type/body expression attached to that declaration

File: bitloops/src/adapters/languages/go/edges.rs (hunk 3 above)
Comment (threeguts) [reply bitloops/src/adapters/languages/go/edges.rs:393]: Golang specific struct embeddings

File: bitloops/src/adapters/languages/go/canonical.rs (hunk 1 above)
Comment (threeguts) [reply bitloops/src/adapters/languages/go/canonical.rs]: Okay, on it!

File: bitloops/src/adapters/languages/go/edges.rs (hunk 4 above)
Comment (threeguts) [reply bitloops/src/adapters/languages/go/edges.rs]: I'll look into it

File: bitloops/src/adapters/languages/go/canonical.rs (hunk 2 above)
Comment (giorgosnty) [reply bitloops/src/adapters/languages/go/canonical.rs:25]: Ok understood. Not sure if  need the body of the type but let's keep it for now

File: bitloops/src/host/language_adapter/kinds.rs [hunk 5]
```diff
@@ -0,0 +1,86 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub(crate) enum LanguageKind {
```
Comment (giorgosnty) [inline bitloops/src/host/language_adapter/kinds.rs:2]: Here let's create a specific for each language e.g. GolangKinds, TsJsKinds etc,  not a single that has everything because then it doesn't add a lot of value.

Comment (giorgosnty) [reply bitloops/src/host/language_adapter/kinds.rs:2]: Ideally to be also strings so that you can do Golang::func without having to cast to string

File: bitloops/src/host/devql/tests/devql_tests/extraction_go.rs [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+                && edge.from_symbol_fqn == "service/run.go::Run"
+                && edge.to_target_symbol_fqn.as_deref() == Some("service/run.go::helper")
+        })
+        .count();
+    let helper_reference_edges = edges
+        .iter()
+        .filter(|edge| {
+            edge.edge_kind == EdgeKind::References
+                && edge.from_symbol_fqn == "service/run.go::Run"
+                && (edge.to_target_symbol_fqn.as_deref() == Some("service/run.go::helper")
```
Comment (giorgosnty) [inline bitloops/src/host/devql/tests/devql_tests/extraction_go.rs:272]: Why do we use "||" ? Which one should happen? Not clear what we are testing

File: bitloops/src/host/devql/tests/devql_tests/extraction_go.rs [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+                    || edge.to_symbol_ref.as_deref() == Some("package::service::helper"))
+        })
+        .count();
+
+    assert_eq!(helper_call_edges, 1);
+    assert_eq!(helper_reference_edges, 0);
+}
+
+#[test]
+fn extract_go_qualified_type_references_do_not_duplicate_single_imported_type_usage() {
```
Comment (giorgosnty) [inline bitloops/src/host/devql/tests/devql_tests/extraction_go.rs:282]: Again not sure what duplicate means here? we are checking we are not having 2 reference edges?

Comment (giorgosnty) [reply bitloops/src/host/devql/tests/devql_tests/extraction_go.rs:282]: or because it says context and Context?

File: bitloops/src/adapters/languages/ts_js/canonical.rs [hunk 8]
```diff
@@ -1,66 +1,66 @@
 use crate::host::devql::CanonicalKindProjection;
-use crate::host::language_adapter::{CanonicalMapping, MappingCondition};
+use crate::host::language_adapter::{CanonicalMapping, LanguageKind, MappingCondition};
 
 pub(crate) static TS_JS_CANONICAL_MAPPINGS: &[CanonicalMapping] = &[
     CanonicalMapping {
-        language_kind: "function_declaration",
+        language_kind: LanguageKind::FunctionDeclaration.as_str(),
```
Comment (giorgosnty) [inline bitloops/src/adapters/languages/ts_js/canonical.rs]: Here we should use the specific enum

File: bitloops/src/host/devql/tests/devql_tests/extraction_go.rs (hunk 6 above)
Comment (threeguts) [reply bitloops/src/host/devql/tests/devql_tests/extraction_go.rs:272]: It uses OR because it might misclassify helper() and represent it either as a local reference or an unresolved package reference while we want it to classify only as a Call edge

File: bitloops/src/host/devql/tests/devql_tests/extraction_go.rs (hunk 7 above)
Comment (threeguts) [reply bitloops/src/host/devql/tests/devql_tests/extraction_go.rs:282]: Yes , it checks that one imported type usage produces exactly one reference edge

File: bitloops/src/host/devql/tests/devql_tests/extraction_go.rs (hunk 6 above)
Comment (giorgosnty) [reply bitloops/src/host/devql/tests/devql_tests/extraction_go.rs:272]: got it

**Review by giorgosnty — APPROVED** [review]

**Review by giorgosnty — APPROVED** [review]
